### PR TITLE
feat(gatsby): Support redirects in serve

### LINKED ISSUE #19441: Gatsby serve should return 301/302s for redirects
## Summary

We've been running a broken link checker as part of our site's CI process (it runs gatsby build, gatsby serve in the background, and then scans localhost:9000 for broken links). When I upgraded from Gatsby 1 to 2 the URLs that redirect which ~used to return a 301~ are now returning a 404 :(

(edit: just realised they used to return a directory listing, not a 301 - so this is unrelated to the upgrade but I think this would still be a useful addition!) 

### Basic example

gatsby-node:
```js
    createRedirect({
      fromPath: `old`,
      isPermanent: true,
      redirectInBrowser: true,
      toPath: `new`
    })
```

```bash
yarn global add broken-link-checker
gatsby build
gatsby serve
```

In another terminal:
```bash
blc -r -o http://localhost:9000/
```

> Getting links from: http://localhost:9000/
> ├─BROKEN─ http://localhost:9000/old (HTTP_404)

### Motivation

So we can distinguish between broken links and redirects before the site is deployed

### PR DESCRIPTION
**Pull request by alexlouden**
## Description

`gatsby serve` now returns a 301 or 302 (if `isPermanent` is false) for URLs that match redirects. 
It reads redirects from `.cache/redirects.json`.

I've tested it locally but not sure how to test this, please let me know :)

## Related Issues

Fixes #19441

### CODE REVIEW

File: packages/gatsby/src/commands/serve.js [hunk 1]
```diff
@@ -61,6 +61,38 @@ const matchPathRouter = (matchPaths, options) => (req, res, next) => {
   return next()
 }
 
+const readRedirects = async program => {
+  const filePath = path.join(program.directory, `.cache`, `redirects.json`)
+  let rawJSON = `[]`
+  try {
+    rawJSON = await fs.readFile(filePath)
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/commands/serve.js]: Can't we require the file instead of reading it and then parsing it?

Comment (alexlouden) [reply packages/gatsby/src/commands/serve.js]: Good point! I followed the same pattern that was already in use for reading `match-paths.json`, but I'm happy to change them both over to require if you'd like. 

Do you think there'd be any issues with require if the file is created/updated after serve is first run, or while it's running?

Comment (alexlouden) [reply packages/gatsby/src/commands/serve.js]: Hey @sidharthachatterjee - any thoughts on the above?

**Comment by sidharthachatterjee** [comment]
See https://github.com/gatsbyjs/gatsby/issues/19441#issuecomment-558090794

**Pull request closed without merging**